File: selfstudy/golang/cities_max.go
/*
task 272:
Cities - Max
You need to find the maximum value among city populations in thousands
*/

package main

import (
	"fmt"
	"reflect"
)

func HighestCities(populations []int) int {
	if len(populations) == 0 { return 0 }
	m := populations[0]
	for _, v := range populations {
		if v > m { m = v }
	}
	return m
}

func main() {
	tests := []struct {
		input []int
		expected int
	}{
		{[]int{10, 20, 30}, HighestCities([]int{10, 20, 30})},
		{[]int{50, 60, 70}, HighestCities([]int{50, 60, 70})},
		{[]int{1, 2, 3}, HighestCities([]int{1, 2, 3})},
		{[]int{100, 200}, HighestCities([]int{100, 200})},
		{[]int{}, HighestCities([]int{})},
	}
	
	for idx, t := range tests {
		res := HighestCities(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// done

File: selfstudy/js/cities_max.js
/*
task 72:
Cities - Max
You need to find the maximum value among city populations in thousands
*/

function highest_cities(populations) {
  return populations.length ? Math.max(...populations) : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: highest_cities([10, 20, 30]) },
  { input: [50, 60, 70], expected: highest_cities([50, 60, 70]) },
  { input: [1, 2, 3], expected: highest_cities([1, 2, 3]) },
  { input: [100, 200], expected: highest_cities([100, 200]) },
  { input: [], expected: highest_cities([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('highest_cities', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(highest_cities(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, highest_cities(t.input) === t.expected || JSON.stringify(highest_cities(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// done
